### PR TITLE
add powershell script to set build numbers correctly

### PR DESCRIPTION
**Pull request by rohit21agrawal**
Fixes: https://github.com/NuGet/Client.Engineering/issues/36

### CODE REVIEW

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# This script is used to dynamically set some build variables during VSTS build.
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: PowerShell syntax for documenting scripts is like following
```ps1
<#
.SYNOPSIS
Short description

.DESCRIPTION
Detailed description

.EXAMPLE
.\build.ps1
To run full clean build, e.g after switching branches

.EXAMPLE
.\build.ps1 -f
Fast incremental build
#>
```

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    [string]$BuildInfoJson,
+    [string]$BuildRTM
+)
+
+if ($BuildRTM -eq 'true')
+{
+    Write-Host "##vso[task.setvariable variable=NupkgOutputDir;]ReleaseNupkgs"
+    do{
+        Start-Sleep -s 15
+        Write-Host "Waiting for buildinfo.json to be generated..."
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: Write the message before `Start-Sleep`?

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    until (Test-Path $BuildInfoJson)
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: Need to set a global timeout or number of retries. Forever waiting is not desired.

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+else
+{
+    $revision = Get-Content $BuildCounter
+    $newBuildCounter = [System.Decimal]::Parse($revision)
+    $newBuildCounter++
+    Set-Content $BuildCounter $newBuildCounter
+    Write-Host "##vso[task.setvariable variable=Revision;]$newBuildCounter"
+    $jsonRepresentation = "
+    {
+        `"BuildNumber`":  `"$newBuildCounter`",
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: Write `BuildNumber` in numeric format so you don't have to parse it.

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# This script is used to dynamically set some build variables during VSTS build.
+# Specifically, this script reads the buildcounter.txt file in the $(DropRoot) to
+# determine the build number of the artifacts, also it sets the $(NupkgOutputDir)
+# based on whether $(BuildRTM) is true or false.
+param
+(
+    [string]$BuildCounter,
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: make it numeric `[int]`

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    [string]$BuildInfoJson,
+    [string]$BuildRTM
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: Make this a `[switch]`

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# This script is used to dynamically set some build variables during VSTS build.
+# Specifically, this script reads the buildcounter.txt file in the $(DropRoot) to
+# determine the build number of the artifacts, also it sets the $(NupkgOutputDir)
+# based on whether $(BuildRTM) is true or false.
+param
+(
+    [string]$BuildCounter,
+    [string]$BuildInfoJson,
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: Mark this with `[Parameter(Mandatory=$True)]`.

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    Write-Host "##vso[task.setvariable variable=Revision;]$currentBuild" 
+}
+else
+{
+    $revision = Get-Content $BuildCounter
+    $newBuildCounter = [System.Decimal]::Parse($revision)
+    $newBuildCounter++
+    Set-Content $BuildCounter $newBuildCounter
+    Write-Host "##vso[task.setvariable variable=Revision;]$newBuildCounter"
+    $jsonRepresentation = "
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: This shouldn't be a string. Declare it as 
```ps1
$jsonRepresentation = @{
  BuildNumber: N
}
```

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    $json = (Get-Content $BuildInfoJson -Raw) | ConvertFrom-Json
+    $currentBuild = [System.Decimal]::Parse($json.BuildNumber)
+    Write-Host "##vso[task.setvariable variable=Revision;]$currentBuild" 
+}
+else
+{
+    $revision = Get-Content $BuildCounter
+    $newBuildCounter = [System.Decimal]::Parse($revision)
+    $newBuildCounter++
+    Set-Content $BuildCounter $newBuildCounter
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: `$BuildCounter = $newBuildCounter`

File: scripts/cibuild/SetVstsBuildNumber.ps1 [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    Write-Host "##vso[task.setvariable variable=Revision;]$newBuildCounter"
+    $jsonRepresentation = "
+    {
+        `"BuildNumber`":  `"$newBuildCounter`",
+        `"CommitHash`":  `"$env:BUILD_SOURCEVERSION`",
+        `"BuildBranch`":  `"$env:BUILD_SOURCEBRANCHNAME`"
+    }";
+
+    New-Item $BuildInfoJson -Force
+    $jsonRepresentation | ConvertFrom-Json | ConvertTo-Json | Set-Content $BuildInfoJson
```
Comment (alpaix) [inline scripts/cibuild/SetVstsBuildNumber.ps1]: Fix issues above, so that statement could be simplified.

**Review by alpaix — CHANGES_REQUESTED** [review]
As this script does more stuff than just a build number consider renaming it to smth like `ConfigureVstsBuild.ps1`

File: scripts/cibuild/SetVstsBuildNumber.ps1 (hunk 4 above)
Comment (rohit21agrawal) [reply scripts/cibuild/SetVstsBuildNumber.ps1]: @alpaix this was actually a path to a file, so i have updated the variable names accordingly

File: scripts/cibuild/SetVstsBuildNumber.ps1 (hunk 5 above)
Comment (rohit21agrawal) [reply scripts/cibuild/SetVstsBuildNumber.ps1]: @alpaix it was a path to the file

File: scripts/cibuild/SetVstsBuildNumber.ps1 (hunk 6 above)
Comment (rohit21agrawal) [reply scripts/cibuild/SetVstsBuildNumber.ps1]: @alpaix VSTS limitation for parallel builds, it only passes the string value out of [true,false]

**Comment by rohit21agrawal** [comment]
@alpaix made changes as per feedback

**Review by alpaix — APPROVED** [review]
LGTM

**Pull request merged**